dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1062 @@
+    for(const auto& stencil : stencil_specs.stencilSpecs()) {
+      auto stencil_combination = stencil;
+
+      setRegionSpec(stencil_combination);
+      while (nextRegion(stencil, stencil_combination)){
+        setRegionSpec(stencil_combination);
+      }
+    }
+  }
+
+
+  template<typename... ARGS>
+  HaloSpec(const ARGS&... args) {
+    std::vector<HaloRegionSpecT> tmp{args...};
+    for(auto& spec : tmp) {
+      _specs[spec.index()] = spec;
+      ++_num_regions;
+    }
+  }
+
+  HaloSpec(const self& other){
+    _specs = other._specs;
+  }
+
+  constexpr HaloRegionSpecT spec(const region_index_t index) const {
+    return _specs[index];
+  }
+
+  constexpr extent_t extent(const region_index_t index) const {
+    return _specs[index].extent();
+  }
+
+  constexpr std::size_t numRegions() const{
+    return _num_regions;
+  }
+
+  const SpecsT& haloSpecs() const{
+    return _specs;
+  }
+
+private:
+  void setRegionSpec(const StencilT& stencil) {
+        auto index = HaloRegionSpecT::index(stencil);
+        auto max = stencil.max();
+        if(_specs[index].extent() == 0)
+          ++_num_regions;
+        if(max >_specs[index].extent())
+          _specs[index] = HaloRegionSpecT(index,max);
+  }
+
+  bool nextRegion(const StencilT& stencil, StencilT& stencil_combination) {
+    for(auto d = 0; d <NumDimensions; ++d) {
+      if(stencil[d] == 0)
+        continue;
+      stencil_combination[d] = (stencil_combination[d] == 0) ? stencil[d] : 0;
+      if(stencil_combination[d] == 0) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+private:
+  SpecsT _specs{};
+  std::size_t _num_regions{0};
+}; // HaloSpec
+
+template <typename ElementT, typename PatternT, typename PointerT = GlobPtr<ElementT, PatternT>,
+          typename ReferenceT = GlobRef<ElementT>>
+class RegionIter : public std::iterator<std::random_access_iterator_tag, ElementT,
+                                        typename PatternT::index_type, PointerT, ReferenceT> {
+private:
+  using SelfT     = RegionIter<ElementT, PatternT, PointerT, ReferenceT>;
+  using GlobMemT  = GlobStaticMem<ElementT, dash::allocator::SymmetricAllocator<ElementT>>;
+  using ViewSpecT = typename PatternT::viewspec_type;
+
+  static const dim_t NumDimensions = PatternT::ndim();
+
+public:
+  using index_type = typename PatternT::index_type;
+  using size_type  = typename PatternT::size_type;
+
+  using reference       = ReferenceT;
+  using const_reference = const reference;
+  using pointer         = PointerT;
+  using const_pointer   = const pointer;
+
+public:
+  /**
+   * Constructor, creates a block boundary iterator on multiple boundary
+   * regions.
+   */
+  RegionIter(GlobMemT& globmem, const PatternT& pattern, const ViewSpecT& _region_view,
+             index_type pos, index_type size)
+      : _globmem(globmem), _pattern(pattern), _region_view(_region_view), _idx(pos), _size(size),
+        _max_idx(_size - 1), _myid(dash::myid()), _lbegin(globmem.lbegin()) {}
```

Comment:
The iterator should be team-aware, use `pattern.team.myid()` instead of `dash::myid()`